src/SignalR/server/StackExchangeRedis/src/Internal/RedisProtocol.cs

```diff
@@ -104,101 +108,118 @@ public byte[] WriteAck(int messageId)
 
         public int ReadAck(ReadOnlyMemory<byte> data)
         {
+            var reader = new MessagePackReader(data);
+
             // See WriteAck for format
-            ValidateArraySize(ref data, 1, "Ack");
-            return MessagePackUtil.ReadInt32(ref data);
+            ValidateArraySize(ref reader, 1, "Ack");
+            return reader.ReadInt32();
         }
 
-        private void WriteHubMessage(Stream stream, HubMessage message)
+        private void WriteHubMessage(ref MessagePackWriter writer, HubMessage message)
         {
             // Written as a MessagePack 'map' where the keys are the name of the protocol (as a MessagePack 'str')
             // and the values are the serialized blob (as a MessagePack 'bin').
 
             var serializedHubMessages = _messageSerializer.SerializeMessage(message);
 
-            MessagePackBinary.WriteMapHeader(stream, serializedHubMessages.Count);
+            writer.WriteMapHeader(serializedHubMessages.Count);
 
             foreach (var serializedMessage in serializedHubMessages)
             {
-                MessagePackBinary.WriteString(stream, serializedMessage.ProtocolName);
+                writer.Write(serializedMessage.ProtocolName);
 
                 var isArray = MemoryMarshal.TryGetArray(serializedMessage.Serialized, out var array);
                 Debug.Assert(isArray);
-                MessagePackBinary.WriteBytes(stream, array.Array, array.Offset, array.Count);
+                writer.Write(array);
             }
         }
 
-        public static SerializedHubMessage ReadSerializedHubMessage(ref ReadOnlyMemory<byte> data)
+        public static SerializedHubMessage ReadSerializedHubMessage(ref MessagePackReader reader)
         {
-            var count = MessagePackUtil.ReadMapHeader(ref data);
+            var count = reader.ReadMapHeader();
             var serializations = new SerializedMessage[count];
             for (var i = 0; i < count; i++)
             {
-                var protocol = MessagePackUtil.ReadString(ref data);
-                var serialized = MessagePackUtil.ReadBytes(ref data);
+                var protocol = reader.ReadString();
+
+                /********************* REVIEW : need feedback (Will remove lines / comment before merge) **********************************/
+                // REVIEW : is that ToArray necessary ?
+                // REVIEW : Returning a "ReadOnlySequence<byte>?" have significant code change accross the repository, so for now i limited the impact here
+                // REVIEW : As my understanding is limited, it will allocate only if multiple sequence are not one next to another in memory ?
+                // REVIEW : previous call to "reader.ReadBytes()" was returning a "byte[]"
+                // REVIEW : not sure if the defaulting to "Array.Empty<byte>()" is a good thing either or if it was trhowing before
+
+                var serialized = reader.ReadBytes()?.ToArray() ?? Array.Empty<byte>();
```

Comment:
Yes, you must copy into an array. `ReadBytes()` returns a `ReadOnlySequence<byte>` directly from the sequence being read, and that sequence generally is based on rented memory so you must not expect any of that memory to out-live the reading code path. Copying it into an array gives you memory that you control the lifetime for which matches the old behavior. So this is good.

Reply:
thx